### PR TITLE
add stricter range check for audit feedback

### PR DESCRIPTION
**Pull request by boolafish**
### Note
- Move the check of Merkle proof non empty to `checkMembership` function under `Merkle.sol`
- Add stricter range check in `PosLib`:
   - `txIndex` should be max to 2^16 - 1. Also, use `uint16` for the field.
   - Since there is a limit on ExitPriority that only 54 bits are reserved for `blockNum` + `txIndex`. Set the `MAX_BLOCK_NUM = ((2 ** 54 - 1) - MAX_TX_INDEX) / (BLOCK_OFFSET / TX_OFFSET);` and add check with that range.
- Add tests for edge values

### CODE REVIEW

File: plasma_framework/contracts/src/utils/PosLib.sol [hunk 1]
```diff
@@ -7,17 +7,20 @@ import "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+        uint16 txIndex;
         uint16 outputIndex;
     }
 
     uint256 constant internal BLOCK_OFFSET = 1000000000;
     uint256 constant internal TX_OFFSET = 10000;
-    uint256 constant internal MAX_TX_INDEX = BLOCK_OFFSET / TX_OFFSET;
+    
+    uint256 constant internal MAX_OUTPUT_INDEX = TX_OFFSET - 1;
+    // since we are using merkle tree of depth 16, max tx index size would be 2^16 - 1
```
Comment (pgebal) [inline plasma_framework/contracts/src/utils/PosLib.sol]: would be -> is

File: plasma_framework/test/src/exits/payment/controllers/PaymentChallengeIFENotCanonical.test.js [hunk 2]
```diff
@@ -677,6 +686,21 @@ contract('PaymentChallengeIFENotCanonical', ([_, ifeOwner, inputOwner, outputOwn
...
+                wrongPosition.outputIndex = 123;
+
+                await expectRevert(
+                    this.exitGame.respondToNonCanonicalChallenge(
+                        this.challengeArgs.inFlightTx,
+                        buildUtxoPos(wrongPosition.blockNum, wrongPosition.txIndex, wrongPosition.outputIndex),
+                        this.inFlightTxInclusionProof,
+                        { from: ifeOwner },
+                    ),
+                    'Should only pass in txPos with outputIndex equals to 0',
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/controllers/PaymentChallengeIFENotCanonical.test.js]: Let's shorten that to `Output index of txPos has to be 0`.

File: plasma_framework/test/src/utils/PosLib.test.js [hunk 3]
```diff
@@ -7,90 +7,140 @@ contract('PosLib', () => {
...
+                });
+            });
+        });
+
+        it('should fail when transaction index exceeds uint16 limit', async () => {
+            const txIndexTooLarge = 2 ** 16;
+            const position = this.blockNum * BLOCK_OFFSET + txIndexTooLarge * TX_OFFSET + this.outputIndex;
+            await expectRevert(
+                this.contract.decode(position),
+                'txIndex exceed the size of uint16',
```
Comment (pgebal) [inline plasma_framework/test/src/utils/PosLib.test.js]: exceed -> exceeds

File: plasma_framework/test/src/utils/PosLib.test.js [hunk 4]
```diff
@@ -7,90 +7,140 @@ contract('PosLib', () => {
...
+        });
+
+        it('should fail when block number exceeds max size allowed in PlasmaFramework', async () => {
+            const position = (maxBlockNum.add(new BN(1))).mul(new BN(BLOCK_OFFSET))
+                .add((new BN(0)).mul(new BN(TX_OFFSET)))
+                .add(new BN(0))
+                .toString();
+            await expectRevert(
+                this.contract.decode(position),
+                'blockNum exceed max size allowed in PlasmaFramework',
```
Comment (pgebal) [inline plasma_framework/test/src/utils/PosLib.test.js]: exceed -> exceeds

File: plasma_framework/test/src/utils/PosLib.test.js [hunk 5]
```diff
@@ -7,90 +7,140 @@ contract('PosLib', () => {
...
+            const position = this.blockNum * BLOCK_OFFSET + txIndexTooLarge * TX_OFFSET + this.outputIndex;
+            await expectRevert(
+                this.contract.decode(position),
+                'txIndex exceed the size of uint16',
+            );
+        });
+
+        it('should fail when block number exceeds max size allowed in PlasmaFramework', async () => {
+            const position = (maxBlockNum.add(new BN(1))).mul(new BN(BLOCK_OFFSET))
+                .add((new BN(0)).mul(new BN(TX_OFFSET)))
```
Comment (pgebal) [inline plasma_framework/test/src/utils/PosLib.test.js]: We don't need that line and next one.
